### PR TITLE
Added slider to choose number of tracks

### PR DESCRIPTION
**Pull request by NoamShuster**

### CODE REVIEW

File: src/artist-selection/artist-picker.tsx [hunk 1]
```diff
@@ -50,11 +52,11 @@ export const ArtistPicker: React.FC<ArtistPickerProps> = observer(({spotifyApi,
             }
         </>
     )
-    //TODO NOAM add a slider to choose the amount of songs you wish to get in the created playlist, and pass as limit to getRecommendations
 });
 
 interface ArtistPickerProps {
     spotifyApi: SpotifyApi;
     selectedGenres: string[];
+    onSliderChange: (e: ChangeEvent<HTMLInputElement>) => void;
```
Comment (NitsanAmit) [inline src/artist-selection/artist-picker.tsx:60]: I think "onSliderValueChange" is more descriptive

File: src/artist-selection/artists-slider.css [hunk 2]
```diff
@@ -0,0 +1,38 @@
+.slidecontainer {
```
Comment (NitsanAmit) [inline src/artist-selection/artists-slider.css:1]: Let's move this to emotion styled components together?

File: src/artist-selection/artists-slider.tsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                max="50"
+                step="5"
+                onChange={(e) => onChange(e)}
+            />
+        </div>
+    );
+});
+
+interface ArtistsSliderProps {
+    onSliderChange: (e: ChangeEvent<HTMLInputElement>) => void;
```
Comment (NitsanAmit) [inline src/artist-selection/artists-slider.tsx:33]: See my comment above about the variable name

File: src/artist-selection/artists-slider.tsx [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    return (
+        <div className="slidecontainer">
+            <div style={{"marginBottom": "10px"}}>I want <b>{recommendationLimit}</b> tracks on my playlist</div>
+            <input // TODO: restart slider to 20 instead of 50 - "value" doesn't work
+                className="slider"
+                id="myRange"
+                type="range"
+                min="10"
```
Comment (NitsanAmit) [inline src/artist-selection/artists-slider.tsx:23]: Put these in constants, somewhere outside the component

File: src/artist-selection/artists-slider.tsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    const [recommendationLimit, setRecommendationLimit] = useState<number>(50);
+    // const strings = useContext(LocaleContext);
+
+    const onChange = (e: ChangeEvent<HTMLInputElement>) => {
+        setRecommendationLimit(e.target.valueAsNumber);
+        onSliderChange(e);
+    }
+
+    return (
+        <div className="slidecontainer">
```
Comment (NitsanAmit) [inline src/artist-selection/artists-slider.tsx:17]: Are you sure this outer container is needed? The entire page should already be centered

File: src/artist-selection/artists-slider.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import React, {ChangeEvent, useContext, useState} from "react";
+import {observer} from "mobx-react";
+import {LocaleContext} from "../spotifun";
+import "./artists-slider.css"
+
+export const ArtistsSlider: React.FunctionComponent<ArtistsSliderProps> = observer(({onSliderChange}) => {
+
+    const [recommendationLimit, setRecommendationLimit] = useState<number>(50);
```
Comment (NitsanAmit) [inline src/artist-selection/artists-slider.tsx:8]: You can try using this value to control the initial slider value. Look up 'React controlled components'

File: src/networking/networking.helper.ts [hunk 7]
```diff
@@ -36,7 +37,7 @@ export const convertUserToAppObject = (user: userDTO): User => {
     return {
         id: user.id,
         name: user.display_name,
-        image: user.images[0].url,
+        image: user.images?.length ? user.images[0].url : happy,
```
Comment (NitsanAmit) [inline src/networking/networking.helper.ts:40]: 💯

File: src/networking/spotify.api.ts [hunk 8]
```diff
@@ -52,19 +52,19 @@ export class SpotifyApi {
             });
     }
 
-    getRecommendations = (genres?: string[], artists?: string[], retry = false): Promise<Track[]> => {
+    getRecommendations = (limit: number, genres?: string[], artists?: string[], retry = false): Promise<Track[]> => {
```
Comment (NitsanAmit) [inline src/networking/spotify.api.ts:55]: I think it's better to make the new limit param optional, and append it to the end of the list.. you can give it a default value (20)

File: src/networking/spotify.api.ts [hunk 9]
```diff
@@ -52,19 +52,19 @@ export class SpotifyApi {
...
+        return fetch(`${baseUrl}?${queryParams}&limit=${limit}`, this.getHeaders())
             .then(this.readResponse)
             .then(response => response.tracks)
             .then((tracks) => convertTracksToAppObject(tracks))
             .catch(() => {
                 if (retry) {
                     return Promise.reject("Error getting recommendations");
                 }
-                return this.onTokenExpired().then(() => this.getRecommendations(genres, artists, true));
+                return this.onTokenExpired().then(() => this.getRecommendations(limit, genres, artists, true));
```
Comment (NitsanAmit) [inline src/networking/spotify.api.ts:67]: The limit param should be the third / last param of the recommendations function IMO

File: src/playlist-creator.tsx [hunk 10]
```diff
@@ -16,6 +16,7 @@ export const PlaylistCreator: React.FunctionComponent<{ authService: AuthService
     const [user, setUser] = useState<User>();
     const [selectedGenres, setSelectedGenres] = useState<string[]>([]);
     const [selectedArtists, setSelectedArtists] = useState<string[]>([]);
+    const [recommendationLimit, setRecommendationLimit] = useState<number>(20);
```
Comment (NitsanAmit) [inline src/playlist-creator.tsx:19]: Why store this value both here and inside the slider component? You can pass it as a prop to the slider

File: src/playlist-review/playlist-review-store.ts [hunk 11]
```diff
@@ -13,12 +13,13 @@ export class PlaylistReviewStore {
...
 
-    constructor(private readonly spotifyApi: SpotifyApi, private readonly selectedGenres: string[], private readonly selectedArtists: string[]) {
+    constructor(private readonly spotifyApi: SpotifyApi, private readonly selectedGenres: string[],
+                private readonly selectedArtists: string[], private readonly playlistLimit: number) {
         this.init();
     }
 
     private init() {
-        this.spotifyApi?.getRecommendations(this.selectedGenres, this.selectedArtists)
+        this.spotifyApi?.getRecommendations(this.playlistLimit, this.selectedGenres, this.selectedArtists)
```
Comment (NitsanAmit) [inline src/playlist-review/playlist-review-store.ts:22]: See my comment above about params order

File: src/playlist-review/playlist-review-styles.tsx [hunk 12]
```diff
@@ -119,4 +119,5 @@ export const AlbumImageContainer = styled.div`
 export const StyledHeartButton = styled.img`
     height: 24px;
     filter: invert(74%) sepia(60%) saturate(4555%) hue-rotate(318deg) brightness(109%) contrast(114%);
+    cursor: pointer;
```
Comment (NitsanAmit) [inline src/playlist-review/playlist-review-styles.tsx:122]: liga🥇

File: src/playlist-review/playlist-review.tsx [hunk 13]
```diff
@@ -48,4 +48,5 @@ interface PlaylistReviewProps {
     userId: string;
     selectedGenres: string[];
     selectedArtists: string[];
+    playlistLimit: number;
```
Comment (NitsanAmit) [inline src/playlist-review/playlist-review.tsx:51]: maybe 'tracksLimit' would be clearer?

File: src/artist-selection/artist-picker.tsx (hunk 1 above)
Comment (NoamShuster) [reply src/artist-selection/artist-picker.tsx:60]: Agree

File: src/artist-selection/artists-slider.tsx (hunk 5 above)
Comment (NoamShuster) [reply src/artist-selection/artists-slider.tsx:17]: It's not centered without it

File: src/playlist-review/playlist-review-store.ts (hunk 11 above)
Comment (NoamShuster) [reply src/playlist-review/playlist-review-store.ts:22]: If I move it to the end I have another default value "retry" - what should I do in this situation?

**Pull request merged**